tests/testthat/test-utility-extra.R

```diff
@@ -0,0 +1,400 @@
+  "Intersecting cones works (when no intersection)", {
+    ref = m4ma::iCones_r(p1, 175, p2[-n, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, 175, p2[-n, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one cone intersecting and seen)", {
+    ref = m4ma::iCones_r(p1, a_not_front, p2[3, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, a_not_front, p2[3, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one cone intersecting but not seen)", {
+    ref = m4ma::iCones_r(p1, a_not_front,
+                         p2[3, , drop = FALSE], r, objects_occlude)
+    res = m4ma::iCones_rcpp(p1, a_not_front,
+                            p2[3, , drop = FALSE], r, objects_occlude)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when all pedestrians are occluded in front)", {
+    ref = m4ma::iCones_r(p1, a_front,
+                         p2[-n, , drop = FALSE], r, objects_occlude)
+    res = m4ma::iCones_rcpp(p1, a_front,
+                            p2[-n, , drop = FALSE], r, objects_occlude)
+    testthat::expect_equal(res, ref)
+  })
+
+testthat::test_that("Intersecting cones to cells transformation works", {
+  # Make test independent of iCones
+  cones = c(rnorm(6))
+  names(cones) = as.character(1:6)
+  v = 0.5
+  ref = m4ma::iCones2Cells_r(cones, v)
+  res = m4ma::iCones2Cells_rcpp(cones, v)
+  testthat::expect_equal(res, ref)
+})
+
+testthat::test_that(
+  "Blocked angle works (when intersecting cones and seen)", {
+    ref = m4ma::blockedAngle_r(n, state, p_pred, objects)
+    res = m4ma::blockedAngle_rcpp(n-1, state, p_pred, objects)
+    testthat::expect_equal(res, if (length(ref) == 0) unname(ref) else ref)
+  }
+)
+
+testthat::test_that(
+  "Blocked angle works (when no intersecting cones)", {
+    a_back = c(175, 175, 175)
+    names(a_back) = nms
+    state_occluded = state
+    state_occluded$a = a_back
+    ref = m4ma::blockedAngle_r(n, state_occluded, p2, objects_occlude)
+    res = m4ma::blockedAngle_rcpp(n-1, state_occluded, p_pred, objects_occlude)
+    testthat::expect_equal(res, if (length(ref) == 0) unname(ref) else ref)
+    testthat::expect_equal(res, numeric(0))
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when seen and in front)", {
+    state_leaders = state
+    ref = m4ma::getLeaders_r(n, state_leaders, centres, objects)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when not seen and in front)", {
+    state_leaders = state
+    ref = m4ma::getLeaders_r(n, state_leaders, centres, objects_occlude)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres, objects_occlude)

```

Comment:
consider the opportunity to hide the '-1' operation within your c++ function (it really doesn't affect the whole logic at all). This way you give less room for bugs and errors to your user which may not remember that he has to provide a different n than the R function.

Reply:
Good point, I will do that!